Persist spks derived from `KeychainTxOutIndex`

evanlinjin opened the pull request:
Replaces #1960
Fixes #1953
Fixes #1964 

### Description

Users with large wallet and/or complex descriptors may experience slow startup of `KeychainTxOutIndex`. This PR addresses this problem by providing the option to persist derived spks so that they no longer need to be re-derived on startup.

The `IndexedTxGraph` API has been changed for better ergonomics.

Compared to #1960, this is a more longterm solution that does not depend on multi-threading logic.

### Changelog notice

```md
Changed
  - `KeychainTxOutIndex::new` to take in an additional parameter `persist_spks` to control whether derived spks are cached and persisted across restarts. The default of `persist_spks` is false.
  - `KeychainTxOutIndex` methods (`lookahead_to_target, `next_unused_spk`, `reveal_next_spk`) now return changesets as they may derive spks to be persisted.
  - The `InsertDescriptorError` type now wraps descriptors in `Box` to reduce enum size.

Added
  - `spk_cache` field to `indexer::keychain_txout::ChangeSet` which persists derived spks.
  - `IndexedTxGraph::from_changeset` - allows constructing from `indexed_tx_graph::ChangeSet` and only indexing when ready.
  - `IndexedTxGraph::reindex` method.

Fixed
  - `KeychainTxOutIndex::reveal_to_target` so that it actually returns `None` if the `keychain` does not exist.
```

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo +nightly fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
feat(chain)!: Persist spks derived from `KeychainTxOutIndex`

evanlinjin committed:
fix(chain)!: API and logical issues in `KeychainTxOutIndex`

Thanks to @valuedmammal for the suggestions

evanlinjin committed:
feat(chain): `KeychainTxOutIndex`: Debug build checks

* When merging changesets, assert that spk of a given descriptor id &
  derivation index does not get changed.
* When reading spk from cache, check the spk by deriving it.

evanlinjin committed:
feat(chain): Add `KeychainTxOutIndex::from_changeset` constructor

This incentivies constructing `KeychainTxOutIndex` from a changeset
before inserting descriptors (to make use of the spk cache).

evanlinjin committed:
feat(chain): `KeychainTxOutIndex`: Make spk cache optional

Also added staging changes to `ChangeSet::spk_cache`. This way, we can
avoid returning `ChangeSet`s for `apply_changeset` and
`insert_descriptor`.

* `KeychainTxOutIndex::new` now takes in an additional `use_spk_cache`
  parameter.
* Fixed `reveal_to_target` method to actually return `None` if the
  keychain does not exist.

evanlinjin committed:
feat(chain)!: Clean up ergonomics of `IndexedTxGraph`

* `new` is now intended to construct a fresh indexed-tx-graph
* `from_changeset` is added for constructing indexed-tx-graph from a
  previously persisted changeset
* added `reindex` for calling after indexer mutations that require it
* reintroduce `Default` impl

evanlinjin committed:
docs(chain): Adds docs for persisting spks in `KeychainTxOutIndex`

ValuedMammal commented on crates/chain/src/rusqlite_impl.rs:
```diff
@@ -533,10 +535,28 @@ impl keychain_txout::ChangeSet {
+            descriptor_id TEXT NOT NULL REFERENCES {}, \
+            spk_index INTEGER NOT NULL, \
+            spk BLOB NOT NULL, \
+            PRIMARY KEY (descriptor_id, index) \
```
Small typo in column name.

```suggestion
            PRIMARY KEY (descriptor_id, spk_index) \
```

ValuedMammal commented on crates/chain/src/rusqlite_impl.rs:
```diff
@@ -533,10 +535,28 @@ impl keychain_txout::ChangeSet {
+    pub fn schema_v1() -> String {
+        format!(
+            "CREATE TABLE {} ( \
+            descriptor_id TEXT NOT NULL REFERENCES {}, \
```
While testing I ran into a failed foreign key constraint because bdk attempted to write to the derived spks table before any scripts were revealed hence the last_revealed table was empty.
```suggestion
            descriptor_id TEXT NOT NULL, \
```

ValuedMammal committed:
fix(rusqlite_impl): Fix derived spks create table statement

Do not reference last revealed table, in case none are revealed.
Correct SQL column name.

ValuedMammal reviewed (approved):
ACK 62767f0702e6497d08fea702b747fd6e7fa72b8c

notmandatory commented:
~~Might have a problem with the sqlite store. I'm manually testing and looks like it's storing the same script for all indexes. I'll keep looking but wanted to give a headsup not to merge this yet.~~
Nevermind, it's working fine, my sqlite select just wan't showing the blog data.

notmandatory reviewed (approved):
ACK 62767f0702e6497d08fea702b747fd6e7fa72b8c

tested with `bdk-cli` with cache enabled and sqlite schema updated and new table is being populated.

ValuedMammal commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -79,6 +73,87 @@ impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I>
+    ///
+    /// let (graph, reindex_cs) =
+    ///     IndexedTxGraph::from_changeset(persisted_changeset, move |idx_cs| -> anyhow::Result<_> {
+    ///         // e.g. KeychainTxOutIndex needs descriptors that weren’t in its CS
```
```rs
// e.g. KeychainTxOutIndex needs descriptors that weren’t in its change set.
```

ValuedMammal commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -186,24 +200,44 @@ impl<K: Clone + Ord + Debug> Indexer for KeychainTxOutIndex<K> {
 
 impl<K> KeychainTxOutIndex<K> {
-    /// Construct a [`KeychainTxOutIndex`] with the given `lookahead`.
+    /// Construct a [`KeychainTxOutIndex`] with the given `lookahead` and `use_spk_cache` boolean.
```
```rs
/// Construct a [`KeychainTxOutIndex`] with the given `lookahead` and `persist_spks` boolean.
```

ValuedMammal commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -186,24 +200,44 @@ impl<K: Clone + Ord + Debug> Indexer for KeychainTxOutIndex<K> {
+    /// // Derive 20 future addresses per chain and persist + reload script pubkeys via ChangeSets:
+    /// let idx = KeychainTxOutIndex::<&'static str>::new(20, true);
+    ///
+    /// // Derive 10 future addresses per chain without persistence:
```
Per "keychain" ?

ValuedMammal commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -215,6 +249,64 @@ impl<K> KeychainTxOutIndex<K> {
 impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+    /// Construct `KeychainTxOutIndex<K>` from the given `changeset`.
+    ///
+    /// Shorthand for called [`new`] and then [`apply_changeset`].
```
Shorthand for "calling"...

ValuedMammal commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -834,14 +998,22 @@ impl<K: core::fmt::Debug> core::fmt::Display for InsertDescriptorError<K> {
+///
+/// It tracks:
+/// 1. `last_revealed`: the highest derivation index revealed per descriptor.
+/// 2. `spks`: the cache of derived script pubkeys to persist across runs.
```
```rs
/// 2. `spk_cache`: ...
```
